File: 35-search-insert-position/35-search-insert-position.java
class Solution {
    public int searchInsert(int[] nums, int target) {
        int first = 0;
        int last = nums.length - 1;
        int mid = 0; 
        
        while(first <= last){
            mid = first + (last-first)/2;
            
            if(nums[mid] == target) return mid;
            
            else if(nums[mid] > target) {
                if(mid == 0 || nums[mid-1] < target) return mid;
                last = mid - 1;
            }
            
            else if(nums[mid] < target) {
                if(mid == nums.length-1 || nums[mid+1] > target) return mid+1;
                first = mid + 1;
            }
            
        }
        
        
        return mid;
    }
}

File: 35-search-insert-position/35-search-insert-position.py
class Solution:
    def searchInsert(self, nums: List[int], target: int) -> int:
        left = 0
        right = len(nums)-1
        mid = 0
        
        while left <= right:
            mid = (left+right)//2
            if nums[mid] == target: return mid
            
            elif target < nums[mid]:
                if mid != 0 and nums[mid-1] == target: return mid-1
                else: right = mid-1
                    
            else:
                if mid!=len(nums)-1 and nums[mid+1] == target: return mid+1
                else: left = mid+1
                    
        if left>right: mid = left
        return mid
      